### PR TITLE
Add new usage reports

### PR DESCRIPTION
**Pull request by AudriusButkevicius**
This incorporates some of @imsodin's work.

# **THIS CANNOT BE DEPLOYED EASILY AS IT DOES AN ALTER THAT WILL LOCK SHIT BADLY**

So add the columns ourselves first somehow.

Screenshot:
ss

### CODE REVIEW

**Comment by AudriusButkevicius** [comment]
# **THIS CANNOT BE DEPLOYED EASILY AS IT DOES AN ALTER THAT WILL LOCK SHIT BADLY**

**Comment by AudriusButkevicius** [comment]
This can be even further optimized in the template, and we can remove not null from the alter to make the migration faster.

**Comment by calmh** [comment]
Looks awesome for the topmost few meters of screenshot, then

32518880-d5da7a98-c402-11e7-871e-dcc726faf38c

**Comment by AudriusButkevicius** [comment]
Different colours per group is close to impossible, as you can't iterate a map (k, v) and get an index for the iteration. Re booleans, there was a guy working on multiple listener support so let's leave it as it is, on the hopes that it will happen.

**Comment by calmh** [comment]
Hardly impossible. Here's some random typing into my browser:

```go
var colors = []string{"red", "green", "blue", "olive", "banana", ...}
var colorMap = make(map[string]int)
func keyToColor(key string) string {
  idx, ok := colorMap[key]
  if !ok {
    idx = len(colorMap) % len(colors)
    colorMap[key] = idx
  }
  return colors[idx]
}
```

Inject that with a funcmap. In the template

```
<span style="background-color: {{key | keyToColor}};">...</span>
```

As for there being multiple GUI listeners that's fine, but that doesn't make TLS enablement less of a boolean, or do I misunderstand you? This is just like the folder features above for example.

**Comment by AudriusButkevicius** [comment]
All group features are no longer like:

```
if rep.DeviceUses.DynamicAddr > 0 {
    features["Device"]["v2"]["Addresses, dynamic"]++
}
``` 

but more like:

```
features["Device"]["v2"]["Addresses, dynamic"] += rep.DeviceUses.DynamicAddr
``` 

In a general context, it's still percentages for options that are on or off, but we show the count as well, which might be much more than devices reporting. Sadly only the progress bar bullshit has tooltips to add meaning to that. I could change that I guess.

**Comment by calmh** [comment]
Hmm yeah ok. I was thinking that the group bars at the bottom were for things where the whole was made up of more than one category and not on/off things. That is, if we have

`Mode, send-only: 25%`

then why do we have

`TLS: [Enabled.... | Disabled]`

In my mind these are both things that are well represented by a percentage of how many reported instances are on vs off.

**Comment by imsodin** [comment]
I think the grouped features should be displayed in a single column, to allow for more labels to be displayed instead of just being accessible via tooltips.

As for grouped vs absolute display: I think the two display different types of data, so it would be justified to show certain features in both ways. As far as I can see you didn't change anything in "my" upper part, so what is in there and what is not is yet pretty random (as I didn't think about grouped features).

**Comment by AudriusButkevicius** [comment]
Anyways, this is not finished.
@calmh is right that TLS enabled disabled makes no sense, I can easily move it to the other section.
I simply shoved it under GUI, as I didn't want to have 2 GUI sections (one with percentage, one with bars for things that are not binary).

**Comment by AudriusButkevicius** [comment]
As for bar widths, I think small is ok, as there is a summary tooltip when you hover on the section label that shows all percentages (even 0% which doesn't have a hoverable area in the bar).

File: cmd/ursrv/main.go [hunk 1]
```diff
@@ -897,87 +924,217 @@ func getReport(db *sql.DB) map[string]interface{} {
...
+
+			if rep.RestartOnWakeup {
+				features["various"]["v3"]["Restart after suspend"]++
+			}
+
+			if rep.CustomStunServers {
+				features["various"]["v3"]["Custom stun servers"]++
+			}
+
+			if rep.UpgradeAllowedPre {
```
Comment (imsodin) [inline cmd/ursrv/main.go]: This is are already in v2 reports, do you count them again for v3 on purpose?

Comment (AudriusButkevicius) [reply cmd/ursrv/main.go]: Must be a copy paste thing

**Comment by AudriusButkevicius** [comment]
Some examples where bars makes more sense than percentages.

```
rep.DeviceUses.DynamicAddr
rep.DeviceUses.StaticAddr
```

These are counts, so having both static and dynamic on any of my devices, adds +1 on each.
Yet across all of my devices, I might have 99 dynamic and 1 static, in which case shifting the balance.

Yes, this does not make sense for GUI.

**Comment by imsodin** [comment]
And at the same time it might also be interesting to know the percentage of instances that use static addresses at all. There is even unused space in the percentage device features table area, so you could have your cake and eat it too.

**Comment by AudriusButkevicius** [comment]
New version.
**# This can be merged as I did the db migration**

image

**Comment by AudriusButkevicius** [comment]
Alternative version pulling in other stats:

image

**Comment by imsodin** [comment]
I just realized that  percentages are now relative to all devices/folders instead of to reports. This gives more weight to big clusters (based on the unproven argument that if you a user uses a feature on one folder/device, it's more likely that it will also be used on other devices), but I think it makes sense. However then the total number of devices/folders should be given directly above the respective lists.

Alternative version doesn't live in this PR, right? Looks cool :)

I would combine the system ram with the memory usage graph (and consistently use RAM, memory or both together) and also combine folder size with number of files. And the hashing performance could be combined with cpu cores.

Data transfer per day needs a legend and I would place it below the metrics and version&co tables, I would consider those stats more relevant.

**Comment by AudriusButkevicius** [comment]
Data transfer doesn't need a legend as it has one as you hover.

**Comment by AudriusButkevicius** [comment]
If @calmh is happy, I'll merge, deploy and fix block stats later.

**Comment by calmh** [comment]
I’m happy. That thing super ducking cool.

**Comment by calmh** [comment]
Hah, autocorrect because mobile. I mean’t “fucking” of course.

**Comment by AudriusButkevicius** [comment]
@calmh fine for performance to be on front page?

**Comment by calmh** [comment]
Yeah I don't remember why i hid it, if it was for performance reasons (expensive to generate?) or due to lack of energy do describe what they showed or what it was.

On 9 Nov 2017, at 16:50, Audrius Butkevicius <notifications@github.com<mailto:notifications@github.com>> wrote:


@calmh<https://github.com/calmh> fine for performance to be on front page?

**Pull request merged**